File: algorithm/acm_whu/c/noahanimal.c
#include<stdio.h>
#include<string.h>

int main(){
    int i,j;
    int tmp1,tmp2;
    int spt[100001];
    int space[100],point[100];
    int animal_num, total_space;
    int total_points;
    while(scanf("%d",&animal_num) != EOF){
        for(i=0;i<animal_num;i++){
            scanf("%d%d",&space[i],&point[i]);
        }
        scanf("%d",&total_space);
        memset(spt, 0, sizeof(spt));
        for(i=0;i<animal_num;i++){
            for(j=total_space;j>=0;j--){
                if(j >= space[i]){
                	tmp1 = spt[j];
                	tmp2 = spt[j - space[i]] + point[i];
                    spt[j] = tmp1 > tmp2? tmp1 : tmp2;
                }
            }
        }
        total_points = spt[total_space];
        printf("%d\n",total_points);
    }
    return 0;
}

File: algorithm/acm_whu/python/noahanimal.py
class ani():
    def __init__(self,space,point):
        self.space = space
        self.point = point

def calculate():
    for i in range(0, ta):
        for j in range(ts, -1, -1):
            if j >= al[i].space:
                tmp1 = pl[j]
                tmp2 = pl[j - al[i].space] + al[i].point
                if tmp2 > tmp1:
                    pl[j] = tmp2


pl = [0 for i in range(100001)]
while True:
    try:
        ta = int(input())
    except:
        break
    al = []
    for i in range(ta):
        space,point = [int(i) for i in input().split()]
        a = ani(space, point)
        al.append(a)
    ts = int(input())
    
    for i in range(100001):
        pl[i] = 0
    calculate()
    total_point = pl[ts]
    print(total_point)